graphql/util.go

```diff
@@ -235,6 +236,88 @@ func SchedulePatch(ctx context.Context, patchId string, version *model.Version,
 	}
 }
 
+type PatchProjectVariantsAndTasks struct {
+	Variants []BuildVariantUI
+	Tasks    []string
+}
+
+type BuildVariantUI struct {
+	Name        string
+	DisplayName string
+	Tasks       []string
+}
+
+func getPatchProjectVariantsAndTasks(patchedConfig string, patchProject string) (*PatchProjectVariantsAndTasks, error) {
+	project := &model.Project{}
+	if _, err := model.LoadProjectInto([]byte(patchedConfig), patchProject, project); err != nil {
+		return nil, errors.Errorf("Error unmarshaling project config: %v", err)
+	}
+
+	// retrieve tasks and variant mappings' names
+	variantMappings := []BuildVariantUI{}
+	for _, variant := range project.BuildVariants {
+		variantUI := BuildVariantUI{}
+		variantUI.Name = variant.Name
+		variantUI.DisplayName = variant.DisplayName
+
+		tasksForVariant := []string{}
+		for _, TaskFromVariant := range variant.Tasks {
+			if TaskFromVariant.IsGroup {
+				groupTasks := model.CreateTasksFromGroup(TaskFromVariant, project)
+				for _, t := range groupTasks {
+					tasksForVariant = append(tasksForVariant, t.Name)
+				}
+			} else {
+				tasksForVariant = append(tasksForVariant, TaskFromVariant.Name)
+			}
+		}
+		variantUI.Tasks = tasksForVariant
+		variantMappings = append(variantMappings, variantUI)
+	}
+
+	tasksList := []string{}
+	for _, task := range project.Tasks {
+		// add a task name to the list if it's patchable
+		if !(task.Patchable != nil && !*task.Patchable) {
+			tasksList = append(tasksList, task.Name)
+		}
+	}
+	p := PatchProjectVariantsAndTasks{
+		Variants: variantMappings,
+		Tasks:    tasksList,
+	}
+	return &p, nil
+}
+
+// GetPatchProjectVariantsAndTasksForUI gets the variants and tasks for a project for a patch id
+func GetPatchProjectVariantsAndTasksForUI(ctx context.Context, patchId string) (*PatchProject, error) {
+	patch, err := patch.FindOne(patch.ById(patch.NewId(patchId)))
```

Comment:
We can avoid making this db query if we pass in an APIPatch instead of patchId to this function and add the PatchedConfig field to APIPatch

Reply:
i like it